<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>exolocator: 12_exon_seq_store Namespace Reference</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />

<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">exolocator
   &#160;<span id="projectnumber">1.01</span>
   </div>
   <div id="projectbrief">assembling protein sequences from exons up</div>
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.8.0 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="namespaces.html"><span>Namespace&#160;List</span></a></li>
      <li><a href="namespacemembers.html"><span>Namespace&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div>
<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">12_exon_seq_store Namespace Reference</div>  </div>
</div><!--header-->
<div class="contents">
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:aa683a56ce5b02b6140dcff8940a5fd7e"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#aa683a56ce5b02b6140dcff8940a5fd7e">get_primary_seq_info</a></td></tr>
<tr class="memitem:a7e07147d33945c4af6b7ad2abaf3cd85"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#a7e07147d33945c4af6b7ad2abaf3cd85">get_alt_seq_info</a></td></tr>
<tr class="memitem:a6e2645c20b8ecb2f7644276caa025214"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#a6e2645c20b8ecb2f7644276caa025214">extract_gene_seq</a></td></tr>
<tr class="memitem:aba2efa8f46ef665d66b5f51a0deac1a0"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#aba2efa8f46ef665d66b5f51a0deac1a0">strip_stop</a></td></tr>
<tr class="memitem:a0b65b4fd52a3291d2086528cc1f47858"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#a0b65b4fd52a3291d2086528cc1f47858">canonical_transl_info</a></td></tr>
<tr class="memitem:acf44dc16c6b297ff8db54e486dfd0fba"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#acf44dc16c6b297ff8db54e486dfd0fba">get_canonical_transl</a></td></tr>
<tr class="memitem:a9573a058a9f0bde8a68724ce31cc3093"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#a9573a058a9f0bde8a68724ce31cc3093">get_canonical_exons</a></td></tr>
<tr class="memitem:a163f10dd9974fa59e5dcb0702b722de6"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#a163f10dd9974fa59e5dcb0702b722de6">transl_reconstruct</a></td></tr>
<tr class="memdesc:a163f10dd9974fa59e5dcb0702b722de6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given tha dna sequence, gene information an the list of its canonical exons, reconstruct canonical translation.  <a href="#a163f10dd9974fa59e5dcb0702b722de6"></a><br/></td></tr>
<tr class="memitem:acf9f17f2aa3d0f54e3662c145a606a32"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#acf9f17f2aa3d0f54e3662c145a606a32">compare_seqs</a></td></tr>
<tr class="memdesc:acf9f17f2aa3d0f54e3662c145a606a32"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return true if the two input sequences differ in no more than two positions.  <a href="#acf9f17f2aa3d0f54e3662c145a606a32"></a><br/></td></tr>
<tr class="memitem:a0a266b0c76b80904008c1ac963e9d8d4"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#a0a266b0c76b80904008c1ac963e9d8d4">get_gene_seq</a></td></tr>
<tr class="memdesc:a0a266b0c76b80904008c1ac963e9d8d4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given gene_id, return dna region which reproduces the correct canonical translation.  <a href="#a0a266b0c76b80904008c1ac963e9d8d4"></a><br/></td></tr>
<tr class="memitem:abf226158714d028315bb8905415f81f7"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#abf226158714d028315bb8905415f81f7">get_exon_seqs</a></td></tr>
<tr class="memdesc:abf226158714d028315bb8905415f81f7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return dna sequences belonging to each exon in the input list, according to Ensembl.  <a href="#abf226158714d028315bb8905415f81f7"></a><br/></td></tr>
<tr class="memitem:a6c9dc3a2f058425a6a487fc8b868d658"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#a6c9dc3a2f058425a6a487fc8b868d658">store</a></td></tr>
<tr class="memdesc:a6c9dc3a2f058425a6a487fc8b868d658"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calls <a class="el" href="namespaceel__utils_1_1mysql.html#ad1dbce5e2a628225a3aee9f4bb002d4b">store_or_update()</a> from el_utils.mysql.py: It sets the fixed fields to be exon_id, is_known, and is_sw; The rest of the exon info is updateable.  <a href="#a6c9dc3a2f058425a6a487fc8b868d658"></a><br/></td></tr>
<tr class="memitem:a60046448e48836673785b159fbe34282"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#a60046448e48836673785b159fbe34282">store_exon_seqs</a></td></tr>
<tr class="memdesc:a60046448e48836673785b159fbe34282"><td class="mdescLeft">&#160;</td><td class="mdescRight">The core of the operation: for each exon find and store dna sequence; for canonical exons also add the translation.  <a href="#a60046448e48836673785b159fbe34282"></a><br/></td></tr>
<tr class="memitem:a8517b3ef4d3826fd5ed3ee3b1bf8d539"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespace12__exon__seq__store.html#a8517b3ef4d3826fd5ed3ee3b1bf8d539">main</a></td></tr>
<tr class="memdesc:a8517b3ef4d3826fd5ed3ee3b1bf8d539"><td class="mdescLeft">&#160;</td><td class="mdescRight">Main entry point, but in reality does nothing except taking care of the parallelization.  <a href="#a8517b3ef4d3826fd5ed3ee3b1bf8d539"></a><br/></td></tr>
</table>
<hr/><h2>Function Documentation</h2>
<a class="anchor" id="a0b65b4fd52a3291d2086528cc1f47858"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.canonical_transl_info </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cursor</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>gene_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

</div>
</div>
<a class="anchor" id="acf9f17f2aa3d0f54e3662c145a606a32"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.compare_seqs </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>canonical_translation</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>translated_seq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>verbose</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Return true if the two input sequences differ in no more than two positions. </p>

</div>
</div>
<a class="anchor" id="a6e2645c20b8ecb2f7644276caa025214"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.extract_gene_seq </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>acg</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>seq_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>file_names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>seq_region_strand</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>seq_region_start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>seq_region_end</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

</div>
</div>
<a class="anchor" id="a7e07147d33945c4af6b7ad2abaf3cd85"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.get_alt_seq_info </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cursor</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>gene_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

</div>
</div>
<a class="anchor" id="a9573a058a9f0bde8a68724ce31cc3093"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.get_canonical_exons </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cursor</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>gene_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

</div>
</div>
<a class="anchor" id="acf44dc16c6b297ff8db54e486dfd0fba"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.get_canonical_transl </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>acg</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cursor</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>gene_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

</div>
</div>
<a class="anchor" id="abf226158714d028315bb8905415f81f7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.get_exon_seqs </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>gene_seq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>exons</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Return dna sequences belonging to each exon in the input list, according to Ensembl. </p>

</div>
</div>
<a class="anchor" id="a0a266b0c76b80904008c1ac963e9d8d4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.get_gene_seq </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>acg</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cursor</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>gene_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Given gene_id, return dna region which reproduces the correct canonical translation. </p>

</div>
</div>
<a class="anchor" id="aa683a56ce5b02b6140dcff8940a5fd7e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.get_primary_seq_info </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cursor</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>gene_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

</div>
</div>
<a class="anchor" id="a8517b3ef4d3826fd5ed3ee3b1bf8d539"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.main </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Main entry point, but in reality does nothing except taking care of the parallelization. </p>
<p>The parallelization here is per-species. </p>

</div>
</div>
<a class="anchor" id="a6c9dc3a2f058425a6a487fc8b868d658"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.store </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cursor</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>exons</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>exon_seq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>left_flank</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>right_flank</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>canonical_exon_pepseq</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Calls <a class="el" href="namespaceel__utils_1_1mysql.html#ad1dbce5e2a628225a3aee9f4bb002d4b">store_or_update()</a> from el_utils.mysql.py: It sets the fixed fields to be exon_id, is_known, and is_sw; The rest of the exon info is updateable. </p>

</div>
</div>
<a class="anchor" id="a60046448e48836673785b159fbe34282"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.store_exon_seqs </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>species_list</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>db_info</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>The core of the operation: for each exon find and store dna sequence; for canonical exons also add the translation. </p>
<p>For each species retrieves all protein coding genes, and for each of the genes its full sequence and the list of the known exons; assigns dna sequence to each exon (+ the translation for the canonical exons) and stores it in db.</p>
<p>The reason why only the canonical exons get the translation at this point is that the canonical translation available in Ensembl is used to make sure we are looking at the right genome region. Namely, some sequences have a "patch" with a corrected or completed sequence. The info about that can be found in the assembly_exception table. </p>

</div>
</div>
<a class="anchor" id="aba2efa8f46ef665d66b5f51a0deac1a0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.strip_stop </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pepseq</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

</div>
</div>
<a class="anchor" id="a163f10dd9974fa59e5dcb0702b722de6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def 12_exon_seq_store.transl_reconstruct </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cursor</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>gene_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>gene_seq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>canonical_coding_exons</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>is_mitochondrial</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>verbose</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Given tha dna sequence, gene information an the list of its canonical exons, reconstruct canonical translation. </p>
<p>Pay attention to whether the gene is mitochondrial. Here in particular we can catch the 'false stop codons' corresponding to selenocysteines. (Because they are stored by their position in the translation.) </p>

</div>
</div>
</div><!-- contents -->


<hr class="footer"/><address class="footer"><small>
Generated on Thu Dec 6 2012 18:32:44 for exolocator by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.0
</small></address>

</body>
</html>
